src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -207,21 +207,82 @@ impl LocationState {
+    /// It returns `true` iff `last_recorded` is stronger than `happening_now`, where
+    /// `None` is weakest and `Some(Write)` is strongest.
+    pub fn foreign_access_requires_update(
+        last_recorded: Option<AccessKind>,
+        happening_now: Option<AccessKind>,
+    ) -> bool {
+        match (last_recorded, happening_now) {
+            // if the last access here was a child access, we need to do nothing
+            (None, _) => false,
+            // if the last access here was a foreign access, but our new child does not
+            // survive any foreign access, we must let the next foreign access through again
+            (_, None) => true,
+            // last access was a read, the child survives reads -> no change needed
+            (Some(AccessKind::Read), Some(AccessKind::Read)) => false,
+            // last access was a read, the child survives writes -> no change needed
+            (Some(AccessKind::Read), Some(AccessKind::Write)) => false,
+            // last access was a write, the child survives writes -> no change needed
+            (Some(AccessKind::Write), Some(AccessKind::Write)) => false,
+            // last access was a write, the child does not survive writes -> reset it to read
+            (Some(AccessKind::Write), Some(AccessKind::Read)) => true,
+        }
+    }
+
+    /// Restores the "children are stronger" invariant of `latest_foreign_access` after a retag.
+    /// See also `foreign_access_requires_update`.
+    /// Retags can violate the invariant on `latest_foreign_access`,
+    /// which states that a node's children survive accesses at least as strong as that node itself.
+    /// While retags act as a read for the offsets specified in the retag, it is also present at the other
+    /// offsets, but marked as "lazy." Such lazy nodes can, however, still be affected by foreign accesses.
+    /// This would violate the invariants that children always survive accesses at least as strong as the last
+    /// recorded foreign access at the parent.
+    /// To restore the invariant, we need to weaken the information stored in the parent, until it is weak enough
+    /// to encompass the newly added lazy node.
+    /// This function updates the `latest_foreign_access` for this node, by weakening it if necessary.
+    /// It returns `true` if such a weakening was necessary. If this method returns `true`, then this node's parent
+    /// also needs to be updated. If it returns `false`, then the invariant is restored globally, since all of `self`'s
+    /// parents are already weaker than `self` itself.
+    fn reset_last_foreign_access_after_retag(
+        &mut self,
+        strongest_survivable: Option<AccessKind>,
+    ) -> bool {
+        let needs_update =
+            Self::foreign_access_requires_update(self.latest_foreign_access, strongest_survivable);
+        if needs_update {
+            self.latest_foreign_access = strongest_survivable;
```

Comment:
Might be worth a debug assertion that `strongest_survivable` is "less than" `latest_foreign_access`.

Reply:
I don't understand this suggestion. What do you want to assert, and where? Note that the whole point of this `if` is to figure out whether `strongest_survivable` is "less than" `latest_foreign_access`. So if you `debug_assert` it in the wrong place, you get an assertion failure, and if you `debug_assert` it in the right place, you're just writing code of shape

```
if foo {
  debug_assert!(foo);
```

Reply:
I think I am asking for a `Ord` implementation on `Option<AccessKind>` and then this entire thing can be written much simpler as `self.latest_foreign_access = self.latest_foreign_access.min(strongest_survivable)`.

Reply:
The `Ord` on `Option` seems to be such that `None < Some(_)`, so if you implement `Ord` on `AccessKind` correctly this should just work.

Reply:
I actually used that at one point in time, but removed it because I figured you'd not like it :laughing: